Tdl 18089 make child streams independent of parent stream

prijendev opened the pull request:
# Description of change
- Made child stream independent of parent stream.
- Now, the bookmark for the child stream will also be written as per the parent replication key.
- In the next sync, it will collect only those parent ids which respect the child stream bookmark.
- Unitest case is written for the parent-child bookmark mechanism

**Note**: Removed non-standard test cases as we have separate https://github.com/singer-io/tap-twitter-ads/pull/20 for standard tap tester.
# Manual QA steps
 - Verify that a proper bookmark is written for the child stream.
 - Verify that the child stream respects its own bookmark instead of the parent stream bookmark.
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

dbshah1212 commented on tap_twitter_ads/sync.py:
```diff
@@ -262,6 +263,9 @@ def sync_endpoint(client,
         cursor = get_resource(stream_name, client, path, new_params)
 
+        # Cursor for children to retrieve parent_ids
+        cursor_child = copy.deepcopy(cursor)
```
Why do we need to copy this? I can't see any changes for the cursor.

prijendev replied:
Here, the `cursor` is `generator(yield)`. First, it will be iterated for the parent stream with the parent's bookmark. But, for the child also we want to iterate through all parent records based on the child bookmark and collect `parent_ids`. That's why we are making a `cursor` copy before the parent iteration.

prijendev commented on tests/test_configuration.py:
```diff
@@ -1,87 +0,0 @@
-def config():
```
Removed test cases as we have separate https://github.com/singer-io/tap-twitter-ads/pull/20 for standard tap tester.

savan-chovatiya commented on tap_twitter_ads/streams.py:
```diff
@@ -403,6 +407,59 @@ def sync_endpoint(self,
+
+                        # Bookmark for child stream
+                        child_last_datetime = self.get_bookmark(state, child_stream_name, start_date)
+                        if not child_last_datetime or child_last_datetime is None:
```
```suggestion
                        if not child_last_datetime :
```

savan-chovatiya commented on tap_twitter_ads/streams.py:
```diff
@@ -403,6 +407,59 @@ def sync_endpoint(self,
+                        for record in cursor_child:
+                            # Get dictionary for record
+                            record_dict = self.obj_to_dict(record)
+                            if not record_dict:
+                                # Finish looping
+                                LOGGER.info('Stream: {} - Finished Looping, no more data'.format(stream_name))
+                                break
```
Do we need this? The `for` loop will execute till `cursor_child` have data so I think no need to break the loop specifically.

prijendev replied:
No, I have verified that removing `break` will not cause any issue. So, removed it.

prijendev commented on tap_twitter_ads/streams.py:
```diff
@@ -403,6 +407,59 @@ def sync_endpoint(self,
+                        # Bookmark for child stream
+                        child_last_datetime = self.get_bookmark(state, child_stream_name, start_date)
+                        if not child_last_datetime or child_last_datetime is None:
+                            child_last_datetime = start_date
```
Updated.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -302,82 +308,80 @@ def sync_endpoint(self,
-                                bookmark_value = datetime.strptime(
-                                    record_dict.get(bookmark_field), datetime_format)
-                            # Other bookmarked endpoints use normal UTC format
+                if stream_name in selected_streams:
```
Add comment for this condition.

prijendev replied:
Added comment.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -403,6 +407,53 @@ def sync_endpoint(self,
+                            if bookmark_field:
+                                bookmark_value_str = record_dict.get(bookmark_field)
+                                if bookmark_value_str:
+                                        child_bookmark_value = strptime_to_utc(record_dict.get(bookmark_field))
```
Seems like extra space left.

prijendev replied:
Removed extra space.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -202,7 +203,8 @@ def sync_endpoint(self,
+                    selected_streams=[]):
         
         # endpoint_config variables
         path = hasattr(endpoint_config, 'path') and endpoint_config.path
```
```suggestion
        path = getattr(endpoint_config, 'path', None)
```

prijendev replied:
Updated.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -292,6 +295,9 @@ def sync_endpoint(self,
             cursor = self.get_resource(stream_name, client, path, new_params)
 
+            # Cursor for children to retrieve parent_ids
+            cursor_child = copy.deepcopy(cursor)
```
Don't iterate records twice, In case of parent and child both are selected, use a minimum of parent's bookmark and child's bookmark.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -308,6 +311,11 @@ def sync_endpoint(self,
+            # cursor is an object like a generator(yield). First, it will be iterated for the parent stream with 
+            # the parent's bookmark. But, for the child also we want to iterate through all parent records 
+            # based on the child bookmark and collect parent_ids. That's why we are making a cursor copy before the parent iteration.
+            cursor_child = copy.deepcopy(cursor) # Cursor for children to retrieve parent_ids
```
We can't be able to make a copy of the generator so will the copy work for the cursor?

prijendev replied:
Here, the cursor is not an exact generator. cursor is just an object that is available in the [twitter-sdk](https://github.com/twitterdev/twitter-python-ads-sdk/blob/master/twitter_ads/cursor.py). That's why we can make copy of cursor and we have verified that it is working totally fine.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -318,82 +326,82 @@ def sync_endpoint(self,
+                                # Other bookmarked endpoints use normal UTC format
+                                else:
+                                    bookmark_value = strptime_to_utc(record_dict.get(bookmark_field))
+                                # If first record, set max_bookmark_value
```
```suggestion
                                # If first record then set it as max_bookmark_value
```

prijendev replied:
Updated comment.

dbshah1212 commented on tap_twitter_ads/streams.py:
```diff
@@ -318,82 +326,82 @@ def sync_endpoint(self,
+                                    stream_name, bookmark_field, record_dict))
+                                # pylint: enable=line-too-long
+                                bookmark_value = last_dttm
+                            if bookmark_value < last_dttm:
```
Add a comment to explain why we are breaking the loop.

prijendev replied:
Added comment.

prijendev commented on tap_twitter_ads/streams.py:
```diff
@@ -419,6 +427,53 @@ def sync_endpoint(self,
+                            if bookmark_field:
+                                bookmark_value_str = record_dict.get(bookmark_field)
+                                if bookmark_value_str:
+                                    child_bookmark_value = strptime_to_utc(record_dict.get(bookmark_field))
```
Moved duplicated code to the common function `return_bookmark`.

prijendev commented on tap_twitter_ads/streams.py:
```diff
@@ -464,8 +525,8 @@ def sync_endpoint(self,
-        # Update the state with the max_bookmark_value for the stream
-        if bookmark_field:
+        # Update the state with the max_bookmark_value for the stream if stream is selected
+        if bookmark_field  and stream_name in selected_streams:
```
Updated comment.

kspeer825 commented on tests/test_twitter_ads_bookmark.py:
```diff
@@ -1,179 +1,193 @@
+                        for record in second_sync_messages:
+                            second_sync_child_foreign_keys.add(record['line_item_id'])
+
+                        # Verify that all foreign keys in the 2nd sync are available in in 1st sync also. 
+                        self.assertTrue(second_sync_child_foreign_keys.issubset(first_sync_child_foreign_keys))
```
I'm not sure this assertion is proving the relationship described in the comment. Can you instead find an expected list of foreign keys on the child stream that we'd expect to be repicated on sync 2 based based on the parent bookmark?

prijendev replied:
Here we are asserting that we are getting more parent's children in the first sync than in the second sync. We also had this https://github.com/singer-io/tap-twitter-ads/blob/092e55cdc4751d7dde79e96028c5ddf3dd3118a6/tests/test_twitter_ads_bookmark.py#L170-#L173 assertion,
are you expecting anything else?

prijendev replied:
Added assertion to verify following both scenarios,

- Verify that each child record from 1st sync whose parent record has a replication key value greater than or equal to bookmark value is replicated in 2nd sync.
- Verify all parents of the children in 2nd sync follows the bookmark.

RushT007 commented on tap_twitter_ads/streams.py:
```diff
@@ -206,7 +207,35 @@ def remove_hours_local(self, dttm, timezone):
         return new_dttm
-       
+
+    def return_bookmark(self, bookmark_value_str, datetime_format, record_dict, bookmark_field, stream_name, record_counter, last_dttm):
```
What is the difference between return_bookmark and get_bookmark? Can we rename it appropriately?

prijendev replied:
`get_bookmark` function returns the bookmark value that is available in the state whereas `return_bookmark` calculates the maximum bookmark value by comparing the replication key value that is available in the records and returns it. Updated name of `return_bookmark` to `return_maximum_bookmark`.

kspeer825 commented on tests/test_twitter_ads_bookmark.py:
```diff
@@ -146,6 +146,31 @@ def test_run(self):
+                        # Collect child key-properties from 1st sync records based on foreign keys of 2nd sync records
+                        first_sync_pks = set()
+                        for record in first_sync_messages:
+                            if record['line_item_id'] in second_sync_child_foreign_keys:
```
I think this conditional should be checking against the parent stream records, not the child stream. You need a set of primary keys based off the messages that are sent to the target for the `line_items` stream. But only the messages with replication keys greater than or equal to the state you inject prior to the 2nd sync.

prijendev replied:
We have already added an [assertion ](https://github.com/singer-io/tap-twitter-ads/blob/21b0ab8a07517dae50f511c16283cf8ac265473f/tests/test_twitter_ads_bookmark.py#L165-L173)to verify that the parent record's replication key value is greater than the child's bookmark value. In this assertion, we are validating the id of the parent that is available in the records of the child(2nd sync).

kspeer825 commented on tests/test_twitter_ads_bookmark.py:
```diff
@@ -1,179 +1,218 @@
+                                                    "line_items", {}).get('messages', [])
+                                                if record.get('action') == 'upsert']
+
+                        # Verify all parents of the children in 2nd sync follows bookmark
+                        for record in parent_sync_messages:
+                            if record['id'] in second_sync_child_foreign_keys:
+                                self.assertGreaterEqual(record['updated_at'], second_bookmark_value)
```
I think the logic here is incomplete. This assertion is saying "if a child is synced, the parent's replication key value must be greater/equal to the bookmark" which is good. But another assertion needs to be made that  "ANY parent record with a replication key value greater/equal to the bookmark MUST have it's child record synced". I think currently this could miss the case where the tap fails to replicate all  the child records that it should.

prijendev replied:
We have added assertion to verify that all the child records of any parent with a replication key value greater/equal to the bookmark synced. All the parent records can not have child records. That's why we can not directly assert that all the child records of 1st sync parent are equal to 1st sync child record and the same thing for 2nd sync also.

kspeer825 reviewed (commented):
I updated the parent-child assertions in this commit https://github.com/singer-io/tap-twitter-ads/pull/18/commits/43a3f972ef87254d136afd85f1e9e3b1364268e0. I was having a hard time following the test as-is. I gave some more thought to which assertions were crucial. Please check out the change and let me know if you think this makes sense, or make changes if needed.

prijendev commented on tests/test_twitter_ads_bookmark.py:
```diff
@@ -1,179 +1,235 @@
+
+                    # collect information specific to incremental streams from syncs 1 & 2
+                    first_bookmark_value_utc = self.convert_state_to_utc(
+                        first_bookmark_value)
```
Yes, the tap is saving bookmarks in utc format. I have removed `convert_state_to_utc` from the mentioned line.

prijendev replied:
@kspeer825  The changes which you have suggested are more similar to what we tried, we are good with it. We have just updated the bookmark test case to pass the build. Can you please re-review it now?

prijendev commented:
> I updated the parent-child assertions in this commit [43a3f97](https://github.com/singer-io/tap-twitter-ads/commit/43a3f972ef87254d136afd85f1e9e3b1364268e0). I was having a hard time following the test as-is. I gave some more thought to which assertions were crucial. Please check out the change and let me know if you think this makes sense, or make changes if needed.

@kspeer825 The changes which you have suggested are more similar to what we tried, we are good with it. We have just updated the bookmark test case to pass the build. Can you please re-review it now?
